api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -765,6 +778,28 @@ message HttpConnectionManager {
   // If this config is set, the Proxy-Status HTTP response header field is
   // populated. By default, it is not.
   ProxyStatusConfig proxy_status_config = 49;
+
+  // Configuration options for Header Validation (UHV).
+  // UHV is an extensible mechanism for checking validity of HTTP requests as well as providing
+  // normalization for request attributes, such as URI path.
+  // If the typed_header_validation_config is present it overrides the following options:
+  // `normalize_path`, `merge_slashes`, `path_with_escaped_slashes_action`
+  // `http_protocol_options.allow_chunked_length`.
+  //
+  // The default UHV checks the following:
+  // 1. HTTP/1 header map validity according to `RFC 7230 section 3.2<https://datatracker.ietf.org/doc/html/rfc7230#section-3.2>`_
+  // 2. Syntax of HTTP/1 request target URI and response status
+  // 3. HTTP/2 header map validity according to `RFC 7540 section 8.1.2<https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2`_
+  // 4. Syntax of HTTP/2 pseudo headers
+  // 5. HTTP/3 header map validity according to `draft standard <https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34>`_
+  // 6. Syntax of HTTP/3 pseudo headers
+  // 7. Syntax of Content-Length and Transfer-Encoding
+  // 8. Validation of HTTP/1 requests with both Content-Length and Transfer-Encoding headers
+  // 9. Normalization of the URI path according to `Normalization and Comparison <https://datatracker.ietf.org/doc/html/rfc3986#section-6>`_
+  //    without `case normalization <https://datatracker.ietf.org/doc/html/rfc3986#section-6.2.2.1>`_
+  // [#not-implemented-hide:]
+  // [#extension-category: envoy.http.header_validators]
+  config.core.v3.TypedExtensionConfig typed_header_validation_config = 50;
```

Comment:
I notice that in the UHV APIs, some of the settings have been flipped from false to true. I think this makes sense from a security perspective, but what are your thoughts on how this will be rolled out to users? Presumably at some point this will get turned on and if not set, the user will get the default validator.

Given that we can't tell if false in the options means false or not set, it seems like for the false cases we have to take the new defaults which would be true. Will we allow that to be reverted via runtime? It seems like that is the only way to do it?

I just want to make sure we understand the plan here. Can you potentially update all of the comments to cover that switch so we are all on the same page on how it's going to work?

Reply:
Note that we have 3 states for UHV config:

1. typed_header_validation_config not even present. This will occur when users upgrade Envoy to a version with UHV turned on by default without modifying HCM config. In this case the default UHV will use "legacy" HCM options for configuration. As such the existing deployments will not be affected by Envoy's internal switch to UHV. This is covered in the comment for `typed_header_validation_config` in HCM config. (This is similar to what we did for DNS resolver extension)
1. typed_header_validation_config present but empty. This means operator had to update the control plane and I assume evaluated possible differences in behavior. However even so the default settings of UHV match the legacy defaults we used to have in HCM. I.e. path normalization is disabled by default.
1. typed_header_validation_config present and non empty. This is really a variation of the case above. Again I assume that operator have evaluated possible change in behavior.

There could be a case for changing defaults in case 2. I.e. enabling path normalization by default, such that present but empty UHV configuration runs path normalization. What do you think?

Reply:
IMO we should ultimately change the defaults to what we want them to be for better security. One way of doing this would be to potentially make the defaults runtime switchable such that we change them to the new secure defaults, but allow reverting to the previous defaults until someone supplies a typed_config with them switched off for real?

Reply:
We can take this path as well. We will need to advertise it really well though, as this is a major breaking change to the protocol behavior.

Reply:
I have updated the default behavior as we have discussed.